### PR TITLE
add Blob.Splice()

### PR DESCRIPTION
**Pull request by ahl**
attn @arv

### CODE REVIEW

**Comment by arv** [comment]
add to whitelist

**Comment by arv** [comment]
LGTM

Do you want to do the JS too? Otherwise we should add a bug.

**Comment by ahl** [comment]
I'll take a swing a the js stuff; fingers crossed that babel doesn't mess me up again.

**Comment by ahl** [comment]
@arv can you take a look at the js side when you have a sec?

File: js/src/blob.js [hunk 1]
```diff
@@ -39,6 +41,14 @@ export default class Blob extends Collection<IndexedSequence> {
   get length(): number {
     return this.sequence.numLeaves;
   }
+
+  splice(idx: number, deleteCount: number, ...insert: number[]): Promise<Blob> {
```
Comment (arv) [inline js/src/blob.js]: I'm not sure we want `...insert: number[]` here. Maybe we should pass a `Uint8Array` here instead.

Comment (arv) [inline js/src/blob.js]: `npm version minor` to let npm modules pick up the new API.

**Comment by arv** [comment]
LGTM

@kalman Can you also take a look (TBR)?

File: go/types/blob.go [hunk 2]
```diff
@@ -34,6 +34,27 @@ func (b Blob) Reader() io.ReadSeeker {
 	return &BlobReader{b.seq, cursor, nil, 0}
 }
 
+func (b Blob) Splice(idx uint64, deleteCount uint64, data []byte) Blob {
+	if deleteCount == 0 && len(data) == 0 {
```
Comment (kalman) [inline go/types/blob.go]: why the `&& len(data) == 0`?

Comment (kalman) [inline go/types/blob.go]: wait dumb question never mind

File: go/types/blob.go [hunk 3]
```diff
@@ -34,6 +34,27 @@ func (b Blob) Reader() io.ReadSeeker {
...
+		return b
+	}
+
+	d.Chk.True(idx <= b.Len())
+	d.Chk.True(idx+deleteCount <= b.Len())
+
+	cur := newCursorAtIndex(b.seq, idx)
```
Comment (kalman) [inline go/types/blob.go]: 2 main comments:
1. A more Go-ish interface here would be to implement Splice using the Go `io.*` libraries. I can't think of a good way to do it though :\
2. We should file an issue to make this look like NewStreamingBlob. Boxing/unboxing each byte is super inefficient, which is why that custom chunking code exists. If Splice becomes performance sensitive, especially if `data` becomes large, then we should fix it.

Comment (kalman) [inline go/types/blob.go]: But lgtm.

Comment (kalman) [inline go/types/blob.go]: (it would be good to file that issue then reference it here - note that JS doesn't do the fancy blob writing)

**Pull request merged**